### PR TITLE
Clear garbage collected lambdified functions from the linecache

### LINKED ISSUE #27216: Memory Leak in `sympy.lambdify`
Hi there,

I'm working with an [algorithm](https://github.com/SymposiumOrganization/NeuralSymbolicRegressionThatScales) that relies on calling `sympy.lambdify` hundreds of millions of times (~200M) and noticed the memory usage of the process steadily creeping up and eventually crashing the program.

I was able to trace the problem to the following (see **Demo** below).

Notably, the memory usage still increases despite
1. lambdifying the same equation again
2. "deleting" the lambdified equation

**System Information:**
WSL2 (Windows 10.0.22631 Build 22631)

```
> lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 22.04.3 LTS
Release:        22.04
Codename:       jammy
```
```
> python --version
Python 3.11.10
```
```
> pip freeze
mpmath==1.3.0
psutil==6.1.0
sympy==1.13.3
```

### Demo:

```python
# sympy_memory.py
import argparse
import psutil

import sympy


if __name__ == '__main__':
    # Parse command line arguments
    parser = argparse.ArgumentParser(description='Run equation n times')
    parser.add_argument('-n', type=int, default=1, help='Number of repetitions')
    args = parser.parse_args()

    # Get the current process to monitor memory usage
    process = psutil.Process()

    # Parse an example equation
    equation_str = 'x'
    parsed_equation = sympy.sympify(equation_str)

    print("Memory Usage:")

    for i in range(args.n):
        # Lambdifying the (same) equation in each iteration
        lambdified_equation = sympy.lambdify('x', parsed_equation)

        # Print memory usage every 10000 iterations
        if i % 10000 == 0:
            mem_usage = process.memory_info().rss / 1024 / 1024  # Convert to MB
            print(f"{i}/{args.n}: {mem_usage:.2f} MB")

        # "Deleting" the lambdified equation
        del lambdified_equation
```

```
> python sympy_memory.py -n 100000
Memory Usage:
0/100000: 49.91 MB
10000/100000: 54.29 MB
20000/100000: 59.27 MB
30000/100000: 64.17 MB
40000/100000: 68.04 MB
50000/100000: 72.05 MB
60000/100000: 75.91 MB
70000/100000: 79.78 MB
80000/100000: 83.45 MB
90000/100000: 89.08 MB
```

**Observation:** The memory usage increases by about ~~+390KB~~ (correction: 390B https://github.com/sympy/sympy/issues/27216#issuecomment-2453168997) per lambdified equation, totalling +39MB over 100k lambdifications and deletions.

### Control (to exclude external factors):

```diff
    parsed_equation = sympy.sympify(equation_str)
+    lambdified_equation = sympy.lambdify('x', parsed_equation)

    print("Memory Usage:")

    for i in range(args.n):
-        # Lambdifying the (same) equation in each iteration
-        lambdified_equation = sympy.lambdify('x', parsed_equation)

        # Print memory usage every 10000 iterations
        if i % 10000 == 0:
            mem_usage = process.memory_info().rss / 1024 / 1024  # Convert to MB
            print(f"{i}/{args.n}: {mem_usage:.2f} MB")

-        # "Deleting" the lambdified equation
-        del lambdified_equation
```

```
> python sympy_memory.py -n 100000
Memory Usage:
0/100000: 49.91 MB
10000/100000: 49.91 MB
20000/100000: 49.91 MB
30000/100000: 49.91 MB
40000/100000: 49.91 MB
50000/100000: 49.91 MB
60000/100000: 49.91 MB
70000/100000: 49.91 MB
80000/100000: 49.91 MB
90000/100000: 49.91 MB
```

**Observation:** The memory usage does not increase.

### PR DESCRIPTION
**Pull request by asmeurer**
This fixes an otherwise memory leak that was caused from lambdifying many functions.

Fixes #27216

#### Release Notes

- utilities
  - Fix a memory leak from lambdifying too many functions.

### CODE REVIEW

File: sympy/utilities/lambdify.py [hunk 1]
```diff
@@ -907,8 +908,17 @@ def _lambdifygenerated(x):
     # mtime has to be None or else linecache.checkcache will remove it
     linecache.cache[filename] = (len(funcstr), None, funcstr.splitlines(True), filename) # type: ignore
```
Comment (asmeurer) [inline sympy/utilities/lambdify.py:909]: Fortunately the linecache doesn't actually store any references to the function objects themselves, only to some strings. So I expect this should be robust.

**Comment by oscarbenjamin** [comment]
Is there an advantage in using `weakref.finalize` over `__del__`?

Generally `__del__` is sort of flakey and usually something that I would think you shouldn't rely on. Does `finalize` improve on that in some way?

**Comment by asmeurer** [comment]
The general vibe I get from searching about it is that `weakref.finalize` is more reliable than `__del__`, but I'm not really sure. But more to the point, would it even be possible to use `__del__` here? Could we add it to the generated function object?

**Comment by oscarbenjamin** [comment]
There would have to be a wrapper class to use `__del__` (as suggested in the issue).

The pypy failure is most likely because the test assumes that deleting the object with `del` causes immediate finalisation/free which is not guaranteed with pypy. It would be worth checking if there is still a memory leak with pypy but otherwise we probably don't need a test for this.

**Comment by asmeurer** [comment]
I was wondering how reliable that test would be. I'm seeing if calling `gc.collect()` fixes it. Otherwise I'm happy to delete it.

**Comment by oscarbenjamin** [comment]
Looks good.

**Comment by oscarbenjamin** [comment]
> I'm seeing if calling `gc.collect()` fixes it.

Apparently not since it just failed:
https://github.com/sympy/sympy/actions/runs/11731531902/job/32682693977?pr=26000

**Comment by asmeurer** [comment]
That looks like a failure with the cleanup method itself (the gc.collect was just there to make the test work).

**Comment by asmeurer** [comment]
Anyways, this looks like there's a race condition in the code that looks in the linecache https://github.com/python/cpython/blob/bbe9b21d06c192a616bc1720ec8f7d4ccc16cab8/Lib/linecache.py#L47C5-L47C15. 

Maybe it would be better to replace the entry in the linecache with a blank entry rather than clearing it.

**Comment by asmeurer** [comment]
I also wonder if it would have been better to use `pop` here instead of `if filename in cache: del cache[filename]` here so that it is more atomic. I'm not sure if that makes a difference, though.

**Comment by asmeurer** [comment]
By the way, why is the doctest runner enabling the debugger? That seems like something that shouldn't be done by default, especially on CI, as it could slow down the test runtimes.

**Comment by asmeurer** [comment]
> Anyways, this looks like there's a race condition in the code that looks in the linecache [python/cpython@`bbe9b21`/Lib/linecache.py#L47C5-L47C15](https://github.com/python/cpython/blob/bbe9b21d06c192a616bc1720ec8f7d4ccc16cab8/Lib/linecache.py#L47C5-L47C15).
> 
> Maybe it would be better to replace the entry in the linecache with a blank entry rather than clearing it.

This fix at https://github.com/sympy/sympy/pull/27232

**Comment by asmeurer** [comment]
I guess it is just coming from upstream https://github.com/python/cpython/blob/main/Lib/doctest.py#L1541-L1555. I can't tell if that's actually enabling the debugger or not.

**Comment by asmeurer** [comment]
Maybe we can check the refcount of the cache key before finalizing? How does string interning work in PyPy.

Also, apparently CPython always finalizes an object immediately when its reference count reaches zero, so this issue should only ever occur on PyPy. https://doc.pypy.org/en/latest/cpython_differences.html#differences-related-to-garbage-collection-strategies

**Comment by bjodah** [comment]
Oh I see, so the garbage collector probably ran between these two lines:
- https://github.com/python/cpython/blob/54c63a32d06cb5f07a66245c375eac7d7efb964a/Lib/linecache.py#L52
- https://github.com/python/cpython/blob/54c63a32d06cb5f07a66245c375eac7d7efb964a/Lib/linecache.py#L59
and it's relying on the list of keys it stored in a list remains valid. Checking the refcount of the value wouldn't help then (and I don't think PyPy offers a refcount anyway). I think you're right that an upstream fix is the most robust option.

**Comment by asmeurer** [comment]
Here's the doctest failure for posterity:

```
sympy/series/approximants.py[1] .                                           [OK]
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/runner/work/sympy/sympy/sympy/testing/runtests.py", line 856, in _doctest
    failed = not t.test()
  File "/home/runner/work/sympy/sympy/sympy/testing/runtests.py", line 1407, in test
    self.test_file(f)
  File "/home/runner/work/sympy/sympy/sympy/testing/runtests.py", line 1491, in test_file
    f, t = runner.run(test,
  File "/home/runner/work/sympy/sympy/sympy/testing/runtests.py", line 1862, in run
    self.debugger.reset()
  File "/opt/hostedtoolcache/PyPy/3.9.19/x64/lib/pypy3.9/pdb.py", line 197, in reset
    bdb.Bdb.reset(self)
  File "/opt/hostedtoolcache/PyPy/3.9.19/x64/lib/pypy3.9/bdb.py", line 57, in reset
    linecache.checkcache()
  File "/opt/hostedtoolcache/PyPy/3.9.19/x64/lib/pypy3.9/linecache.py", line 64, in checkcache
    entry = cache[filename]
KeyError: '<lambdifygenerated-12>'
sympy/series/formal.py[15] 
```

The lambdified function itself probably came from some doctest that was run much earlier in the run (nothing in `series` calls lambdify).

**Comment by asmeurer** [comment]
Yeah, I agree ideally we would want `checkcache` itself to clear our entry. Although that function isn't actually called that often (https://github.com/search?q=repo%3Apython%2Fcpython+checkcache+language%3APython&type=code&l=Python), so we would still probably want to call it ourselves in the finalizer to avoid the memleak. The way `checkcache` is written now though that seems impossible. So I don't see how to fix this short of monkeypatching something in `linecache`. 

More generally, better upstream support for dynamically generated linecache entries would be nice. It feels like they're only unofficially supported right now. This feature also gets used by custom interpreters like IPython/Jupyter (although for IPython it's generally not a problem that the cache never gets cleared). 

Just as an aside: just what is the status of PyPy? It seems like they are struggling just to keep up with the oldest supported versions of CPython. I'm not necessarily suggesting we should drop support for it, but I am curious if it will stay relevant in the future.

File: sympy/utilities/lambdify.py [hunk 2]
```diff
@@ -907,8 +908,17 @@ def _lambdifygenerated(x):
...
+    # Remove the entry from the linecache when the object is garbage collected
+    def cleanup_linecache(filename):
+        def _cleanup():
+            if filename in linecache.cache:
+                del linecache.cache[filename]
+        return _cleanup
+
     func = funclocals[funcname]
 
+    weakref.finalize(func, cleanup_linecache(filename))
```
Comment (graingert) [inline sympy/utilities/lambdify.py:920]: I think you want the weakref to be on the `func.__code__`

**Comment by graingert** [comment]
> Also, apparently CPython always finalizes an object immediately when its reference count reaches zero, so this issue should only ever occur on PyPy. 

A function can participate in a reference cycle

> Also, apparently CPython always finalizes an object immediately when its reference count reaches zero, so this issue should only ever occur on PyPy. [doc.pypy.org/en/latest/cpython_differences.html#differences-related-to-garbage-collection-strategies](https://doc.pypy.org/en/latest/cpython_differences.html#differences-related-to-garbage-collection-strategies)

This issue can happen on CPython if the func is referenced by something participating in a reference cycle, eg `funcs = [func]; funcs.append(funcs); del funcs, func`

File: sympy/utilities/lambdify.py (hunk 2 above)
Comment (asmeurer) [reply sympy/utilities/lambdify.py:920]: Can you explain more why that would do anything? I would have thought that would be equivalent.

**Comment by asmeurer** [comment]
> This issue can happen on CPython if the func is referenced by something participating in a reference cycle, eg funcs = [func]; funcs.append(funcs); del funcs, func

That doesn't do anything in my [simple reproducer](https://github.com/pypy/pypy/issues/5109), but I'll take your word for it.

File: sympy/utilities/lambdify.py (hunk 2 above)
Comment (graingert) [reply sympy/utilities/lambdify.py:920]: The function can be deleted, but the code object can outlive it, eg tracebacks references the code object not the function

**Comment by oscarbenjamin** [comment]
Going back to my original comment here:

> Generally `__del__` is sort of flakey and usually something that I would think you shouldn't rely on. Does `finalize` improve on that in some way?

I don't think we should try to rely on deallocation to clean these things up. The right way to manage resources in Python is `with` but if we don't want to ask the user to use `with` then we should at least provide an option that does not require any subsequent cleanup (i.e. not adding to linecache in the first place).

**Comment by graingert** [comment]
I don't think the with will help either, the `__exit__` could run at the same time as linecache.checkcache

**Pull request merged**